primitives/src/signatures/bls_over_bn254.rs

```diff
@@ -0,0 +1,399 @@
+        Hash::hash(&self.sigma, state);
+    }
+}
+
+impl PartialEq for Signature {
+    fn eq(&self, other: &Self) -> bool {
+        self.sigma == other.sigma
+    }
+}
+// =====================================================
+// end of definitions
+// =====================================================
+
+/// Non constant time hash to curve algorithm (a.k.a "hash-and-pray")
+/// The hashing algorithm consists of the following steps:
+///   1. Hash the bytes to a field element `x`.
+///   2. Compute `Y = x^3 + 3`. (Note: the equation of the BN curve is
+/// y^2=x^3+3)   3. Check if `Y` is a quadratic residue (QR), in which case
+/// return `y=sqrt(Y)` otherwise try with `x+1, x+2` etc... until `Y` is a QR.
+///   4. Return `P=(x,y)`
+///
+///  In the future we may switch to a constant time algorithm such as Fouque-Tibouchi <https://www.di.ens.fr/~fouque/pub/latincrypt12.pdf>
+/// * `H` - parameterizable hash function (e.g. SHA256, Keccak)
+/// * `msg` - input message
+/// * `returns` - A group element in G1
+#[allow(non_snake_case)]
+pub fn hash_to_curve<H: Default + DynDigest + Clone>(msg: &[u8]) -> G1Projective {
+    let hasher_init = &[1u8];
+    let hasher = <DefaultFieldHasher<H> as HashToField<BaseField>>::new(hasher_init);
+    let field_elems: Vec<BaseField> = hasher.hash_to_field(msg, 1);
+
+    // General equation of the curve: y^2 = x^3 + ax + b
+    // For BN254 we have a=0 and b=3 so we only use b
+    let coeff_b: BaseField = MontFp!("3");
+
+    let mut x = field_elems[0];
+    let mut Y: BaseField = x * x * x + coeff_b;
+
+    // Loop until we find a quadratic residue
+    while Y.legendre().is_qnr() {
+        // println!("point with x={} is off the curve!!", x_affine);
+        x += BaseField::from(1);
+        Y = x * x * x + coeff_b;
+    }
+
+    // Safe unwrap as `y` is a quadratic residue
+    let y = Y.sqrt().unwrap();
+
+    let g1_affine = G1Affine::new(x, y);
+    G1Projective::from(g1_affine)
+}
+
+impl KeyPair {
+    /// Key-pair generation algorithm
+    pub fn generate<R: Rng>(prng: &mut R) -> KeyPair {
+        let sk = SignKey::generate(prng);
+        let vk = VerKey::from(&sk);
+        KeyPair { sk, vk }
+    }
+
+    /// Key pair generation using a particular sign key secret `sk`
+    pub fn generate_with_sign_key(sk: ScalarField) -> Self {
+        let sk = SignKey(sk);
+        let vk = VerKey::from(&sk);
+        KeyPair { sk, vk }
+    }
+
+    /// Get reference to verification key
+    pub fn ver_key_ref(&self) -> &VerKey {
+        &self.vk
+    }
+
+    /// Get the verification key
+    pub fn ver_key(&self) -> VerKey {
+        self.vk.clone()
+    }
+
+    /// Get the internal of the signing key, namely a P::ScalarField element
+    pub fn sign_key_internal(&self) -> &ScalarField {
+        &self.sk.0
+    }
+
+    /// Signature function
+    pub fn sign<B: AsRef<[u8]>>(&self, msg: &[u8], csid: B) -> Signature {
+        let msg_input = [msg, csid.as_ref()].concat();
```

Comment:
Do we need to hash the message into a fixed-length digest first?

Reply:
As discussed [above](https://github.com/EspressoSystems/jellyfish/pull/223#discussion_r1158635887), I don' t think this is necessary.